### PR TITLE
Testing code with unittest

### PR DESCRIPTION
**Pull request by mamkhari**
This PR does the following:
- Uses unit testing to test  the cases of the two functions(extract_all_sensors
 and splitting_extracted_sensors)

### CODE REVIEW

**Review by mmphego — COMMENTED** [review]
So far everything looks good, some points:
- Add tests for filename and URL related stuff
- You should also test the negative edge case i.e test for brokeness, also consider testing for logs

Check this piece of code for inspiration: https://github.com/ska-sa/katobs/blob/master/katobs/test/subarray_managers/test_subarray_utilities.py#L397+L414 have a chat with @amakhaba about this.

**Comment by mmphego** [comment]
> So far everything looks good, some points:
> 
> * Add tests for filename and URL related stuff
> * You should also test the negative edge case i.e test for brokeness, also consider testing for logs
> 
> Check this piece of code for inspiration: https://github.com/ska-sa/katobs/blob/master/katobs/test/subarray_managers/test_subarray_utilities.py#L397+L414 have a chat with @amakhaba about this.

I meant this function: https://github.com/mamkhari/extract_sensor_names/blob/8209e1460c/extract_unique_sensors_01.py#L33

**Review by mmphego — COMMENTED** [review]
When TDD'ing consider this structure, this will help test edge cases where you test for success, failure and exception raise.

```python

class Test...:
    def test..._success:
        ...
    def test..._failure:
        ...
    def test..._raise_exception:
        ...
```

File: test_extract_unique_sensors_01.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    def test_extract_all_sensors(self):
+        sensor_name = "kat"
+        sensors_data = [
+            "kat.sensor.name_1", "not.kat.sensor.name_1", "kat.sensor.name_2",
+            "kat.sensor.name_3", "ket.name.sensor_2"
+        ]
+        expected_candidate_lines = ['kat.sensor.name_1', 'kat.sensor.name_2',
+                                    'kat.sensor.name_3']
+        self.assertIsInstance(sensor_name, str)
+        self.assertIsInstance(sensors_data, list)
```
Comment (mmphego) [inline test_extract_unique_sensors_01.py]: Test for errors, in the case where someone would pass in a different type.
Check:
```python
with self.assertRaises(<expected exception error>) as err:
    ...
```

Read:
- https://stackoverflow.com/questions/6103825/how-to-properly-use-unit-testings-assertraises-with-nonetype-objects
- https://daniel.perez.sh/blog/2017/unittest-assert-raises/

File: test_extract_unique_sensors_01.py [hunk 2]
```diff
@@ -1,11 +1,38 @@
...
+            mock_request.return_value.status_code = 404
+            self.assertTrue(non_existing_url)
+
+        file_name = "katlego"
+        mock_path = MagicMock()
+        mock_path.Path(file_name).exists().return_value = True
+        self.assertTrue(file_name)
+        self.assertIsInstance(file_name, str)
+        self.assertIsInstance(url, str)
+
```
Comment (mmphego) [inline test_extract_unique_sensors_01.py]: :clap: :clap: Great progress here
@amakhaba great job.

File: test_extract_unique_sensors_01.py [hunk 3]
```diff
@@ -1,11 +1,38 @@
...
+        def Assertlogs(self):
```
Comment (mmphego) [inline test_extract_unique_sensors_01.py]: Not sure why you need a function here. You could just get away with it without a function.

Some sorcery down here, @amakhaba care to explain what's going on?

Comment (amakhaba) [reply test_extract_unique_sensors_01.py]: Hi @mmphego, this is her attempt, i will look at the case you said she should consider then we can have a chat, I havent had time to talk to her

**Comment by amakhaba** [comment]
@mamkhari when @mmphego mentioned the case of trying to test out logs he actually meant
- Testing the behavior when an event either fails, passes and then logs something so, for an example from [line 53-55](https://github.com/mamkhari/extract_sensor_names/blob/master/extract_unique_sensors_01.py#L53-L55), you could test when the case of a file name exist will it log the  message or not.

File: test_extract_unique_sensors_01.py (hunk 3 above)
Comment (mamkhari) [reply test_extract_unique_sensors_01.py]: @mmphego There was no need for a function, I see that now. Thanks!

File: test_extract_unique_sensors_01.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        with self.assertLogs('test_logger', level='DEBUG') as cm:
+            logging.getLogger('test_logger.file_name').debug("file_name exist")
+            logging.getLogger('test_logger.url').debug("URL exist")
+            logging.getLogger('test_logger').error("file_name or url type does exist")
+            self.assertEqual(
+            cm.output, [
+                'DEBUG:test_logger.file_name:file_name exist',
+                'DEBUG:test_logger.url:URL exist',
+                'ERROR:test_logger:file_name or url type does exist']
+            )
```
Comment (mmphego) [inline test_extract_unique_sensors_01.py]: Shouldn't you be testing the `read_sensors` function, as DUT(device under test)?
I noticed that it is not imported, and also you are patching the wrong package instead of `urllib` which was used.

Comment (mamkhari) [reply test_extract_unique_sensors_01.py]: @mmphego I see I was using the wrong package that's not even there on my script. I fixed it, and imported the function that I am testing. Thank you.

File: test_extract_unique_sensors_01.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+from mock import MagicMock
+from mock import patch
+from urllib.request import urlopen
+from extract_unique_sensors_01 import read_sensors
+from extract_unique_sensors_01 import extract_all_sensors
+from extract_unique_sensors_01 import splitting_extracted_sensors
+
+
+class TestExtractSensors(unittest.TestCase):
+    """Test class for function test_read_sensors."""
```
Comment (mmphego) [inline test_extract_unique_sensors_01.py]:
```suggestion
    """Test class for extract_unique_sensors."""
```

File: test_extract_unique_sensors_01.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    def test_read_sensors(self):
+        with patch('urllib.request.urlopen') as mock_urlopen:
+            url = "http://lebogang.com"
+            mock_urlopen.return_value.status_code = 200
+            self.assertTrue(url)
+
+        with patch('urllib.request.urlopen') as mock_urlopen:
+            non_existing_url = "http://google.com/fakeurl"
+            mock_urlopen.return_value.status_code = 404
+            self.assertTrue(non_existing_url)
```
Comment (mmphego) [inline test_extract_unique_sensors_01.py]: @amakhaba Please assist here with the mocks.

You could do something like:
```python
        file_name = MagicMock()
        url = MagicMock()
        DUT = read_sensors(file_name, url)
```

Where DUT is the device under test, and you could interact with it similar to what you did in line #52 and line 64

I cannot think of a better solution for now, but if you need assistance let me know.

File: extract_unique_sensors_01.py [hunk 7]
```diff
@@ -52,14 +52,14 @@ def read_sensors(file_name, url):
...
-        logging.debug("Accesing sensor data from the file_name: %r.", file_name)
+        logging.debug("Accessing sensor data from the file_name: %r.", file_name)
         with open(file_path) as lines:
             sensors_data = lines.readlines()
     elif url:
         try:
             sensors_url = urlopen(url)
             sensors_data = sensors_url.readlines()
-            logging.debug("Accessing sensor data from the url:  %r.", url")
+            logging.debug("Accessing sensor data from the url:  %r.", url)
```
Comment (mmphego) [inline extract_unique_sensors_01.py:61]: Fixed a syntax error, I was hoping that `flake8` or your tests would pick it up.

File: test_extract_unique_sensors_01.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
+import logging
+import unittest
+from pathlib import Path
+from urllib.request import urlopen
+
+from mock import MagicMock, Mock, mock_open, patch
+
+from extract_unique_sensors_01 import (extract_all_sensors, read_sensors,
+                                       splitting_extracted_sensors)
+
```
Comment (mmphego) [inline test_extract_unique_sensors_01.py:10]: I ran [isort](https://github.com/PyCQA/isort) which is a nifty tool for sorting out imports.

Command ran:
```bash
isort -rc .
```

File: test_extract_unique_sensors_01.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            ) as mock_file:
+                with patch.object(logging, "debug") as mocked_logger:
+                    DUT = read_sensors(file_name, url)
+
+        path_exists.assert_called()
+        mock_file.assert_called()
+        mocked_logger.assert_called_with(
+            "Accessing sensor data from the file_name: %r.", file_name
+        )
+        self.assertEqual([read_data], DUT)
```
Comment (mmphego) [inline test_extract_unique_sensors_01.py:32]: Added this example of testing code which you can use to complete `test_read_sensors_from_url`. Hopefully, it's easier to read.
When in doubt reach out - I'll try to explain it.

**Review by mmphego — COMMENTED** [review]
- I added a sample test for testing the `read_sensors` function.
- I also ran [black](https://github.com/psf/black) which auto-formats your code - so that you never worry about [PEP8 Violations](https://www.python.org/dev/peps/pep-0008/)
- Ran [isort](https://github.com/PyCQA/isort) which automagically sorts your modules.

Good luck. I hope the code I implemented will be easier to read...

File: test_extract_unique_sensors_01.py (hunk 9 above)
Comment (mamkhari) [reply test_extract_unique_sensors_01.py:32]: I am not sure if I follow, it's confusing me.
